libcontainer/cgroups/fs/devices.go

```diff
@@ -32,29 +40,63 @@ func (s *DevicesGroup) Set(path string, cgroup *configs.Cgroup) error {
 	}
 
 	devices := cgroup.Resources.Devices
+	oldAllowedDevices, err := readDevicesExceptDefault(path)
+	if err != nil {
+		return err
+	}
+
 	if len(devices) > 0 {
 		for _, dev := range devices {
 			file := "devices.deny"
 			if dev.Allow {
 				file = "devices.allow"
 			}
-			if err := fscommon.WriteFile(path, file, dev.CgroupString()); err != nil {
-				return err
+
+			// For the second time set, we don't deny all devices
+			if dev.Type == defaultDevice.Type && len(oldAllowedDevices) != 0 {
+				file = ""
+			}
+
+			if len(file) > 0 {
+				if err := fscommon.WriteFile(path, file, dev.CgroupString()); err != nil {
+					return err
+				}
 			}
 		}
 		return nil
 	}
 	if cgroup.Resources.AllowAllDevices != nil {
 		if *cgroup.Resources.AllowAllDevices == false {
-			if err := fscommon.WriteFile(path, "devices.deny", "a"); err != nil {
-				return err
+			// For the first time set, we deny all devices to initialize the cgroup
+			if len(oldAllowedDevices) == 0 {
+				if err := fscommon.WriteFile(path, "devices.deny", "a"); err != nil {
+					return err
+				}
 			}
 
-			for _, dev := range cgroup.Resources.AllowedDevices {
-				if err := fscommon.WriteFile(path, "devices.allow", dev.CgroupString()); err != nil {
-					return err
+			newAllowedDevices := make(map[string]Empty)
+			for _, dev := range cgroup.AllowedDevices {
+				newAllowedDevices[dev.CgroupString()] = Empty{}
+			}
```

Comment:
I think this should loop over `cgroup.Devices` given my previous point about how `AllowedDevices` is actually not used by anything. But as I said, most of this logic needs to be torched -- I just wrote a very long block of text about how this patch doesn't handle denying no longer allowed devices, only to realise that this code block actually is used, and in theory does handle that problem...